src/nl/hannahsten/texifyidea/inspections/latex/probablebugs/LatexQuoteInspection.kt

```diff
@@ -0,0 +1,118 @@
+import nl.hannahsten.texifyidea.util.requiredParameter
+import nl.hannahsten.texifyidea.util.toTextRange
+import java.util.regex.Pattern
+import javax.annotation.RegEx
+
+/**
+ * @author Michael Milton
+ */
+class LatexQuoteInspection : TexifyInspectionBase() {
+    override val inspectionGroup = InsightGroup.LATEX;
+    override val inspectionId = "AsciiQuotes";
+
+    /**
+     * Matches anything that looks like a LaTeX quote
+     */
+    private val quotePattern = """["'`]+""".toRegex();
+
+    /**
+     * Defines how to extract a quote pair from a command
+     */
+    data class QuoteMaker(val command: String, val openParam: Int, val closeParam: Int)
+
+    /**
+     * All the ways that csquotes defines a new quotation mark
+     */
+    private val quoteMakers = arrayOf(
+        QuoteMaker("\\MakeOuterQuote", 0, 0),
+        QuoteMaker("\\MakeInnerQuote", 0, 0),
+        QuoteMaker("\\MakeAutoQuote", 0, 1),
+        QuoteMaker("\\MakeForeignQuote", 1, 2),
+        QuoteMaker("\\MakeBlockQuote", 0, 2),
+        QuoteMaker("\\MakeForeignBlockQuote", 1, 3),
+        QuoteMaker("\\MakeHyphenBlockQuote", 1, 3),
+        QuoteMaker("\\MakeHybridBlockQuote", 1, 3),
+    );
+
+    override fun inspectFile(file: PsiFile, manager: InspectionManager, isOntheFly: Boolean): List<ProblemDescriptor> {
+        // Build up a list of acceptable quote pairs. The ith index in both arrays must contain a matching quote
+        // pair. Additional quote pairs are extracted from the csquotes quote making commands
+        val openers = LinkedHashSet(arrayListOf("`", "``"));
+        val closers = LinkedHashSet(arrayListOf("'", "''"));
+        for (quoteMaker in quoteMakers) {
+            for (command in file.commandsInFile(quoteMaker.command)) {
+                val opener = command.requiredParameter(quoteMaker.openParam);
+                val closer = command.requiredParameter(quoteMaker.closeParam);
+
+                if (!opener.isNullOrEmpty() && !closer.isNullOrEmpty()) {
+                    openers.add(opener);
+                    closers.add(opener);
+                }
+            }
+        }
+
+        // Process each of the quote tokens individually
+        val quoteStack = ArrayDeque<Int>();
+        val issues = descriptorList();
+        for (text in findChildrenOfType(file, LatexNormalText::class.java)) {
+            for (match in quotePattern.findAll(text.text)) {
+                val openIndex = openers.indexOf(match.value);
+                if (openIndex > -1) {
+                    // First case: we find a valid opener
+                    // In that case, push the ID of that opener onto the stack
+                    quoteStack.add(openIndex);
+                }
+                else {
+                    val closeIndex = closers.indexOf(match.value);
+                    if (closeIndex == -1) {
+                        // Second case: we find a token that's neither a valid opener nor closer
```

Comment:
If this is the case, I think there is an implementation error ( because all quotes in the regex are either opening or closing quotes) so you should throw an exception

Reply:
No it's not really an error. This case occurs whenever we find a token that looks like a quote, but it's not a "proper quote" that's recognised by latex or has been registered using `csquotes`. For example if the user just typed `"hello"`. The " has not been registered, so it hits this case. Therefore we just want to generate a problem descriptor.

Reply:
I see, in that case I liked the previous description of `Use LaTeX quotes instead of ASCII double quotes` more, because it also tells you (even without the quickfix) what it should be. Plus, it's less wrong in case it is a false positive, as it's phrased as a general guideline instead of telling the user 'this is wrong'.

Reply:
The convention for inspection warnings seems to be a statement like "X should not Y" or "X is not defined" rather than an imperative. So I would rather keep it as is.